src/Mobs/Monster.cpp

```diff
@@ -1248,6 +1258,195 @@ std::unique_ptr<cMonster> cMonster::NewMonsterFromType(eMonsterType a_MobType)
 
 
 
+void cMonster::EngageLoveMode(cMonster *a_Partner)
+{
+	m_LovePartner = a_Partner;
+	m_MatingTimer = 50;  // about 3 seconds of mating
+}
+
+
+
+
+
+void cMonster::ResetLoveMode()
+{
+	m_LovePartner = nullptr;
+	m_LoveTimer = 0;
+	m_MatingTimer = 0;
+	m_LoveCooldown = TPS * 60 * 5;  // 5 minutes
+
+	// when an animal is in love mode, the client only stops sending the hearts if we let them know it's in cooldown, which is done with the "age" metadata
+	m_World->BroadcastEntityMetadata(*this);
+}
+
+
+
+
+
+void cMonster::LoveTick(void)
+{
+	// if we have a partner, mate
+	if (m_LovePartner != nullptr)
+	{
+
+		if (m_MatingTimer > 0)
+		{
+			// If we should still mate, keep bumping into them until baby is made
+			Vector3d Pos = m_LovePartner->GetPosition();
+			MoveToPosition(Pos);
+		}
+		else
+		{
+			// Mating finished. Spawn baby
+			Vector3f Pos = (GetPosition() + m_LovePartner->GetPosition()) * 0.5;
+			UInt32 BabyID = m_World->SpawnMob(Pos.x, Pos.y, Pos.z, GetMobType(), true);
+
+			cMonster * Baby = nullptr;
+
+			m_World->DoWithEntityByID(BabyID, [&](cEntity & a_Entity)
+			{
+				Baby = static_cast<cMonster *>(&a_Entity);
+				return true;
+			});
+
+			if (Baby != nullptr)
+			{
+				Baby->InheritFromParents(this, m_LovePartner);
+			}
+
+			m_World->SpawnExperienceOrb(Pos.x, Pos.y, Pos.z, GetRandomProvider().RandInt(1, 6));
+
+			m_World->DoWithPlayerByUUID(m_Feeder, [&] (cPlayer & a_Player)
+			{
+				a_Player.GetStatManager().AddValue(Statistic::AnimalsBred);
+				if (GetMobType() == eMonsterType::mtCow)
+				{
+					a_Player.AwardAchievement(Statistic::AchBreedCow);
+				}
+				return true;
+			});
+			m_LovePartner->ResetLoveMode();
+			ResetLoveMode();
+		}
+	}
+	else
+	{
+		// We have no partner, so we just chase the player if they have our breeding item
+		cItems FollowedItems;
+		GetFollowedItems(FollowedItems);
+		if (FollowedItems.Size() > 0)
+		{
+			m_World->DoWithNearestPlayer(GetPosition(), static_cast<float>(m_SightDistance), [&](cPlayer & a_Player) -> bool
```

Comment:
Eventually we probably should create a `DoWithNearestPlayers` which would loop the players in order of the distance between the specified point. Now if the nearest player doesn't have the `FollowedItem` equiped, but the next player does he is ignored.